streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/task/CheckpointProcessor.java

```diff
@@ -52,30 +66,32 @@ public class CheckpointProcessor {
 
   @Autowired private SavePointService savePointService;
 
-  public void process(Long appId, CheckPoints checkPoints) {
-    CheckPoints.Latest latest = checkPoints.getLatest();
-    if (latest == null || latest.getCompleted() == null) {
-      return;
-    }
-    CheckPoints.CheckPoint checkPoint = latest.getCompleted();
-    Application application = applicationService.getById(appId);
+  public void process(Application application, @Nonnull CheckPoints checkPoints) {
+    checkPoints.getLatestCheckpoint().forEach(checkPoint -> process(application, checkPoint));
+  }
+
+  private void process(Application application, @Nonnull CheckPoints.CheckPoint checkPoint) {
+    String jobID = application.getJobId();
+    Long appId = application.getId();
     CheckPointStatus status = checkPoint.getCheckPointStatus();
 
     if (CheckPointStatus.COMPLETED.equals(status)) {
-      String cacheId = appId + "_" + application.getJobId();
-      Long latestId =
-          checkPointCache.get(
-              cacheId,
-              key -> {
-                SavePoint savePoint = savePointService.getLatest(appId);
-                return Optional.ofNullable(savePoint).map(SavePoint::getChkId).orElse(null);
-              });
-
-      if (latestId == null || latestId < checkPoint.getId()) {
-        saveSavepoint(checkPoint, application);
+      String cacheId = getCacheIdForCheckpoint(appId, jobID);
+      Long latestChkId = getLatestCheckpointedId(appId, cacheId);
+      if (shouldStoreAsSavepoint(appId, jobID, checkPoint)) {
+        savepointedCache.put(
+            getCacheIdForSavepoint(appId, jobID, checkPoint.getId()), DEFAULT_FLAG_BYTE);
+        saveSavepoint(
+            checkPoint,
+            application.getId(),
+            latestChkId == null || checkPoint.getId() > latestChkId);
+        return;
+      }
+      if (shouldStoreAsCheckpoint(checkPoint, latestChkId)) {
```

Comment:
```suggestion
      if (shouldStoreAsSavepoint(appId, jobID, checkPoint)) {
        savepointedCache.put(
            getCacheIdForSavepoint(appId, jobID, checkPoint.getId()), DEFAULT_FLAG_BYTE);
        saveSavepoint(
            checkPoint,
            application.getId(),
            latestChkId == null || checkPoint.getId() > latestChkId);
        return;
      }
      String cacheId = getCacheIdForCheckpoint(appId, jobID);
      Long latestChkId = getLatestCheckpointedId(appId, cacheId);
      if (shouldStoreAsCheckpoint(checkPoint, latestChkId)) {
```


These two lines should be moved after the savepoint due to they aren't used for savepoint.
```
      String cacheId = getCacheIdForCheckpoint(appId, jobID);
      Long latestChkId = getLatestCheckpointedId(appId, cacheId);
```

Reply:
thx for the comments. let me check it.

Reply:
We still use `latestChkId ` which references the `cacheId`, right?

Reply:
If so, why does savepoint use the latestChkId?

Reply:
We need consider whether the savepoint should be stored as the latest.
I check the `CheckpointProcessor#saveSavepoint` method and maybe there would be a bug  on `latest` value when the chk-n checkpoint completed before chk-n-1 savepoint. 
Please let me know your opinion.

Reply:
I don't understand why savepoint needs the latestChkId.

Actually, I think we don't need the latest field in the database. And the latest can be fill in during qurey checkpoint list. The checkpoint with max triggerTime should be the latest checkpoint.

So I prefer keep old logic that latest is always  true during save chk or savepoint.

Reply:
By the way, the cache still have bug when streampark restarts.

Because the cache will be empty after restart. So I prefer do less here.

Reply:
Thank you so much for the clarify. I got it & updated it.